src/protocols/musig2/mod.rs

```diff
@@ -0,0 +1,167 @@
+    keys: &ExpandedKeyPair,
+    message: Option<&[u8]>,
+    rng: &mut impl Rng,
+) -> PartialNonces {
+    // here we deviate from the spec, by introducing  non-deterministic element (random number)
+    // to the nonce, this is important for MPC implementations
+    let r: [Scalar<Ed25519>; NUMBER_OF_NONCES] = [(); NUMBER_OF_NONCES].map(|_| {
+        Sha512::new()
+            .chain(&[2])
+            .chain(&*keys.expanded_private_key.prefix.to_bytes())
+            .chain(message.unwrap_or(&[]))
+            .chain(rng.gen::<[u8; 32]>())
+            .result_scalar()
+    });
+    let R: [Point<Ed25519>; NUMBER_OF_NONCES] =
+        r.clone().map(|scalar| Point::generator() * &scalar);
+    PartialNonces { r, R }
+}
+
+pub fn partial_sign(
+    nonces_from_other_parties: &[[Point<Ed25519>; NUMBER_OF_NONCES]],
+    my_partial_nonces: PartialNonces,
+    agg_public_key: &PublicKeyAgg,
+    my_keypair: &ExpandedKeyPair,
+    message: &[u8],
+) -> PartialSignature {
+    // Sum up partial nonces from all parties
+    let R: [Point<Ed25519>; NUMBER_OF_NONCES] = nonces_from_other_parties.iter().fold(
+        my_partial_nonces.R,
+        |mut accumulator: [Point<Ed25519>; NUMBER_OF_NONCES],
+         partial_nonce_array: &[Point<Ed25519>; NUMBER_OF_NONCES]| {
+            for (accum_nonce, nonce) in accumulator.iter_mut().zip(partial_nonce_array) {
+                *accum_nonce = &*accum_nonce + nonce;
+            }
+            accumulator
+        },
+    );
+
+    // Compute b as hash of nonces
+    let mut hasher = Sha512::new()
+        .chain(&[3])
+        .chain(&*agg_public_key.agg_public_key.to_bytes(false));
+    for nonce in &R {
+        hasher.update(&*nonce.to_bytes(false));
+    }
+    hasher.update(message);
+    let b: Scalar<Ed25519> = hasher.result_scalar();
+    // Compute effective nonce
+    let (effective_R, effective_r, _) = R[1..].iter().zip(my_partial_nonces.r[1..].iter()).fold(
+        (R[0].clone(), my_partial_nonces.r[0].clone(), b.clone()),
+        |accumulator: (Point<Ed25519>, Scalar<Ed25519>, Scalar<Ed25519>),
+         nonce_tuple: (&Point<Ed25519>, &Scalar<Ed25519>)| {
+            (
+                accumulator.0 + &accumulator.2 * nonce_tuple.0,
+                &accumulator.1 + accumulator.2 * nonce_tuple.1,
+                accumulator.1 * &b,
+            )
+        },
+    );
+    // Compute Fiat-Shamir challenge of signature
+    let sig_challenge = Signature::k(&effective_R, &agg_public_key.agg_public_key, message);
+
+    // Computes the partial signature
+    let partial_signature: Scalar<Ed25519> = sig_challenge
+        * &agg_public_key.musig_coefficient
+        * &my_keypair.expanded_private_key.private_key
+        + effective_r;
+    PartialSignature {
+        R: effective_R,
+        my_partial_s: partial_signature,
+    }
+}
+
+pub fn aggregate_partial_signatures(
+    my_partial_sig: &PartialSignature,
+    partial_sigs_from_other_parties: &[Scalar<Ed25519>],
```

Comment:
This isn't really consistent, I'd instead pass here `&[PartialSignature]` and also check that the Rs all match

Reply:
I fixed everything but this, I don't agree that the partial signatures need to be checked, within the lib as it's not part of the protocol and also makes it slower, if any of them are wrong the final sig will be wrong as well

Reply:
1. This library is slow, very slow compared to what can be done, so I don't think performance is a property we should be talking about.
2. A paper is just a paper, not an implementation guide.